src/List.mo

```diff
@@ -280,12 +338,27 @@ module {
   /// *Runtime and space assumes that `predicate` runs in `O(1)` time and space.
   public func filter<T>(list : List<T>, predicate : T -> Bool) : List<T> {
     let filtered = empty<T>();
-    forEach<T>(
-      list,
-      func(x) {
-        if (predicate(x)) add(filtered, x)
-      }
-    );
+
+    let blocks = list.blocks;
+    let blockCount = blocks.size();
+
+    var i = 1;
+    while (i < blockCount) {
+      let db = blocks[i];
+      let sz = db.size();
+      if (sz == 0) return filtered;
+
+      var j = 0;
+      while (j < sz) {
+        switch (db[j]) {
+          case (?x) if (predicate(x)) add(filtered, x);
+          case null return filtered
+        };
+        j += 1
+      };
+      i += 1
+    };
+
```

Comment:
I would use `forEach` here. It is just a little bit slower (one function call per element). My argument here is that the cost does not matter so much relative to the whole cost. The function is necessarily slow because we have to call `add()` for each individual element that makes it through the filter. Relative to that the single function call will be small. 

This is in the interest of readability, reduces the amount of lines to review for a reviewer.